docs/arch/adr-011-kurl-addon-operator.md

```diff
@@ -0,0 +1,96 @@
+
+## Solution
+
+The operator will be implemented with the following objectives in mind:
+
+- It will be scaffolded using [kubebuilder](https://book.kubebuilder.io/introduction.html).
+- It will be developed using [SOLID](https://en.wikipedia.org/wiki/SOLID) principles and operator best practices.
+- It will distribution agnostic. I.e. it will not be tightly coupled to the kURL Kubernetes distribution.
+- It will manage application add-ons _only_ and _exclude_ managing CRI and CNI resources.
+- It will install and confiure the KOTS admin console.
+- It can be installed on an existing [kURL](https://kurl.sh/) cluster with only Kubernetes, containerd and flannel.
+- The installation mechanism for add-ons will be through the use of Helm Charts.
+- By default the add-on configuration will be opinionated by us but we will enable the end-user to override the configuration for maximum flexibility. Vendors and users will be able to override the configuration using a standard [Helm values](https://helm.sh/docs/chart_template_guide/values_files/) file.
+- It will rely as much as possible on Helm for managing the lifecyle of the add-ons.
+
+### The `Installer` API
+
+The `Installer` API defines a resource for installing specific add-ons.
+
+```yaml
+apiVersion: cluster.kurl.sh/v2beta1
+kind: Installer
+metadata:
+  name: installer-sample
+spec:
+  channel: "1.27"
+  minio:
+    enabled: true
+    configOverride:
+        persistence:
+            enabled: true
+        replicas: 1
+        mode: standalone
+
+```
+
+### The Controllers
+
+The operator will have at minimum one controller which ensures the `Installer` CR is reconciled with the cluster periodically. The `Installer` controller (main controller) will delegate responsiblity for managing add-ons to other controllers. The reconcilation for the main controller will consist of determining the add-ons to be managed from the `Installer` CR and creating a controller for each add-on (addon-on controller).
+
+The add-on controller will be responsible for
+- verifying the requirements for the add-on are met
+- managing the add-on itself
+- ensure add-on is in desired state
+- update the add-on CR status
+
+### Managing the Add-ons
+
+How the add-ons will be managed is still an open question. We have a couple of options:
+
+_Use the Helm Controller provided by Flux_
+
+The [Flux Helm Controller](https://fluxcd.io/flux/components/helm/helmreleases/) affords us the following benefits:
+- Stable [HelmRelease API](https://fluxcd.io/flux/components/helm/api/)
+- Helm release [dependencies](https://fluxcd.io/flux/components/helm/helmreleases/#helmrelease-dependencies)
+- Support for Helm install, upgrade, test, rollback and uninstall
+- Performs Helm uninstall on `HelmRelease` CR removal
+- Chart values override\
```

Comment:
See that with pure Helm we just need to pass the chart values to install/upgrade and we achieve the same, see [here](https://github.com/replicatedhq/brainstorming-ekco/pull/2/files#diff-4d05894b38940f59fcac351cf0cfb3f1ae42eddc423299655677f122dbe4bfe2R169) in the prototype.

So, we could either receive the input from the Installer CR and just pass as map for it, i.e.:

```go
	// Override chart values
	chart.Values = valueOverrides

	// Perform the install or upgrade operation
	if releaseExists(releaseName, actionConfig) {
		_, err = upgradeAction.Run(releaseName, chart)
	} else {
		_, err = installAction.Run(chart, nil)
	}
```

So, I either do not see it as a advantage. Am I overlooking any complexities that Flux handles in these scenarios that I haven't considered?

Reply:
Flux achieves this though composeValues

https://github.com/fluxcd/helm-controller/blob/main/internal/controllers/helmrelease_controller.go#L555-L558

Reply:
This is a good example where the prototype implementation is only a few lines but flux has a much more comprehensive implementation. What's the difference here, is the flux method better and addressing thing we haven't yet addressed or is it more complicated but functionally the same for our use case?

A second question, it looks like in the above prototype code, we replace the chart values with valuesOverrides. Is that true? The intention would be to mergeOverride, IE only override values that are set not replace all values with a new set of values.